nucliadb_vectors/src/formula/mod.rs

```diff
@@ -18,37 +18,41 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 //
 
-use crate::data_point::{Address, DataRetriever};
+use std::collections::HashSet;
 
-#[derive(Debug, Copy, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
-pub enum AtomKind {
-    KeyPrefix,
-    Label,
-}
+use crate::data_point::{Address, DataRetriever};
 
 /// Is a singleton clause.
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
-pub struct AtomClause {
-    kind: AtomKind,
-    value: String,
+pub enum AtomClause {
+    KeyPrefix(String),
+    Label(String),
+    KeyPrefixSet(HashSet<String>),
 }
 impl AtomClause {
-    pub fn new(value: String, kind: AtomKind) -> AtomClause {
-        AtomClause {
-            kind,
-            value,
-        }
-    }
     pub fn label(value: String) -> AtomClause {
-        AtomClause::new(value, AtomKind::Label)
+        Self::Label(value)
     }
     pub fn key_prefix(value: String) -> AtomClause {
-        AtomClause::new(value, AtomKind::KeyPrefix)
+        Self::KeyPrefix(value)
+    }
+    pub fn key_set(set: HashSet<String>) -> AtomClause {
+        Self::KeyPrefixSet(set)
     }
     fn run<D: DataRetriever>(&self, x: Address, retriever: &D) -> bool {
-        match self.kind {
-            AtomKind::KeyPrefix => retriever.get_key(x).starts_with(self.value.as_bytes()),
-            AtomKind::Label => retriever.has_label(x, self.value.as_bytes()),
+        match self {
+            Self::KeyPrefix(value) => retriever.get_key(x).starts_with(value.as_bytes()),
+
+            Self::Label(value) => retriever.has_label(x, value.as_bytes()),
+            Self::KeyPrefixSet(set) => {
+                let key = retriever.get_key(x);
+                let kk = String::from_utf8_lossy(key);
+                let parts: Vec<_> = kk.split('/').collect();
+                if parts.len() < 3 {
+                    return false;
+                }
+                set.contains(&parts[0..3].join("/"))
```

Comment:
Maybe im wrong, but in line `55` you are splitting the whole string by '/' to then take the first three chunks and joining them again by `/`, wouldn't it be better to do this in a single iteration through the key that builds the string you want?

Reply:
Indeed. I just copied this piece of code from the old code at the paragraphs index. I cleaned that one up, but this one remained. I'll make it better.